<div style="border:solid green 2px; padding: 20px">
<b>Гасым, Привет!</b>

Меня зовут Аня Терешонок, и я буду проверять твой проект. Предлагаю общаться на «ты» :) Но если это неудобно - дай знать, и мы перейдем на "вы". 

Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе стать аналитиком данных. Ты уже проделал большую работу над проектом, но давай сделаем его еще лучше. Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их**. Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самой найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> Так выделены самые важные замечания. Без их отработки проект не будет принят. </div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b> Так выделены небольшие замечания. Я надеюсь, что их ты тоже учтешь - твой проект от этого станет только лучше. Но настаивать на их отработке не буду.

</div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> Так я выделяю все остальные комментарии.</div>

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом.** Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>Комментарий студента:</b> Например, вот так.</div>

Всё это поможет выполнить повторную проверку твоего проекта оперативнее. 
    
    
**Общее впечатление от ревью:**
    
Ты проделал большую работу.  Молодец! Давай проработем некоторые комментарии.
    
**Общее впечатление от ревью v2:**
    
Отлично просто доработал проект. Осталось чуть-чуть)
        
**Общее впечатление от ревью v3:**
    
Спасибо за старания! Работу принимаю. Успехов в обучении 😊

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>

Круто, что использовал автоматическое оглавление!

</div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

# План выполнения работы

1.	Загрузите и подготовьте данные. Поясните порядок действий.
2.	Исследуйте баланс классов, обучите модель без учёта дисбаланса. Кратко опишите выводы.
3.	Улучшите качество модели, учитывая дисбаланс классов. Обучите разные модели и найдите лучшую. Кратко опишите выводы.
4.	Проведите финальное тестирование.

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b>

Вступление в работу очень важно. Так ты поясняешь то, чему она посвящена 👍

</div>

## Подготовка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score, roc_auc_score
# import dtale
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from imblearn.over_sampling import SMOTE

<div class="alert alert-success">
<b> Комментарий ревьюера ✔️: </b>
    
Очень хорошо, что вынес все импортируемые библиотеки в одну ячейку. Это делает ноутбук чистым и позволяет избежать ошибок импорта. 
</div>

In [2]:
data = pd.read_csv('/datasets/Churn.csv')

In [3]:
data.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


In [5]:
# Удалим ненужные столбцы
data = data.drop(['RowNumber', 'CustomerId', 'Surname'], axis=1)

In [6]:
# Удалим пропущенные значения
data = data.dropna()

# Конвертируем категориальные переменные в числовые переменные, используя дамми-переменные
data = pd.get_dummies(data, drop_first=True)

# Разделим данные на целевой и обучающие признаки
target = data['Exited']
features = data.drop('Exited', axis=1)

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b>  
    
Как вариант, да, можно просто удалить NaN из датасета, но хорошо было бы добавить небольшое пояснение, почему так тут можно сделать.

</div>

Удаление строк с пропущенными значениями (NaN) из набора данных может быть разумным подходом, если процент пропущенных значений относительно невелик и эти пропущенные значения отсутствуют совершенно случайно. В таких случаях удаление строк с пропущенными значениями не окажет существенного влияния на статистические свойства оставшихся данных и, следовательно, на выводы, которые можно сделать из анализа, как в нашем случае! В этом можно убедиться, посмотрев на изначальный необработанный датасет, где единственный столбец Tenure имел всего лишь 909 пропущенных значений из 10.000.

In [7]:
# dtale.show(data)

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9091 entries, 0 to 9998
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CreditScore        9091 non-null   int64  
 1   Age                9091 non-null   int64  
 2   Tenure             9091 non-null   float64
 3   Balance            9091 non-null   float64
 4   NumOfProducts      9091 non-null   int64  
 5   HasCrCard          9091 non-null   int64  
 6   IsActiveMember     9091 non-null   int64  
 7   EstimatedSalary    9091 non-null   float64
 8   Exited             9091 non-null   int64  
 9   Geography_Germany  9091 non-null   uint8  
 10  Geography_Spain    9091 non-null   uint8  
 11  Gender_Male        9091 non-null   uint8  
dtypes: float64(3), int64(6), uint8(3)
memory usage: 736.9 KB


<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

Отлично! Можно улучшить проект добавить немного больше информации о датасете, проверка на дубли, таблица с кол-вом NaN, также корреляции и графики распредлений.

ℹ️ Дополнительная информация.
    
Часто бывает удобно быстро посмотреть датасет. Для этого существует множество информативных пакетов для быстрого первичного анализа данных. Если интересно, можешь посмотреть, например <a href='https://github.com/man-group/dtale'> это </a> или <a href='https://github.com/ydataai/ydata-profiling'> это </a> 😌.

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Привет, Аня! Спасибо большое, что поделилась своими знаниями о новых для меня инструментах, о которых я впервые слышу)) Начнём с 1-го: я установил D-Tale через pip install dtale (кстати его каждый раз нужно мне переустанавливать на Юпитер нашего тренажера практикума, где я и делают проект - поэтому я закомментил пока что эту часть) и затем импортировал этот пакет, но при использовании функции dtale.show(data) у меня на выходе пустой белый экран(( Не знаю с чем это может быть связано, может ты подскажешь? Спасибо!  </div>

<div class="alert alert-success">
<b>Комментарий ревьюера v3 ✔️:</b> 

Прости, я случайно забыла ответить на этот вопрос в прошлый раз. Попробуй попробовать на локалке или в колабе. В окружении прктикума может не работать. А так, к сожалению, нужно смотреть на ошибку и пробовать исправить. Так прям подскзаать не могу. 
Может <a href="https://stackoverflow.com/questions/60490325/dtale-show-in-jupyter-notebook">тут</a> что-то найдешь. 
</div>

## Исследование задачи

Исследуем баланс классов и потренируем модель логистической регрессии без учета дисбаланса.

In [9]:
print('Количество положительных экземпляров:', target.sum())
print()
print('Количество отрицательных экземпляров:', (target == 0).sum())

Количество положительных экземпляров: 1854

Количество отрицательных экземпляров: 7237


*Набор данных несбалансирован, только 20% положительных экземпляров. Это указывает на то, что положительных примеров (то есть клиентов, которые покинули банк) значительно меньше, чем отрицательных (клиентов, которые остались). Это важно отметить, потому что несбалансированный набор данных может повлиять на производительность некоторых алгоритмов машинного обучения, поскольку они склонны отдавать предпочтение классу большинства.*

In [10]:
# Разделим данные на обучающие, валидационные и тестовые выборки.

# Разделим данные сначало на обучающие и тестовые выборки со стратификацией, чтобы скорректировать смещение прогноза.
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.2, 
                                                                            random_state=12345, stratify=target)

# И затем разделим уже обучающий набор на тренировочный и валидационные выборки со стратификацией.
features_train, features_valid, target_train, target_valid = train_test_split(features_train, target_train, test_size=0.25,
                                                                              random_state=12345, stratify=target_train)

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> 
    
Выборку стоит разделить на 3 части - обучающая, валидационная, тестовая. Обычно делят в соотношении 60:20:20.

</div>


<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b>  
    
В функции `train_test_split` есть интрересный параметр `stratify`. Колонку, которую ты поставишь в качестве значения параметра (обычно, это таргет) будет разделена между выборками таким образом, что соотношение значений в группе в выборках будет одинаковым. Частенько используется при несблансированном таргете для корректировки смещения предсказаний. Можешь прочитать в <a href="https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html#sklearn.model_selection.train_test_split">документации</a>. 😊

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал!  </div>

In [11]:
# Стандартизируем признаки в обоих наборах с помощью StandardScaler(). 
scaler = StandardScaler()
numerical_cols = ['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']
features_train = features_train.copy()
features_valid = features_valid.copy()
features_test = features_test.copy()
features_train[numerical_cols] = scaler.fit_transform(features_train[numerical_cols])
features_valid[numerical_cols] = scaler.transform(features_valid[numerical_cols])
features_test[numerical_cols] = scaler.transform(features_test[numerical_cols])

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

Частая ошибка при стандартизации данных - производить ее на всей выборке. `fit` на трейне и `transform` на трейне и тесте. Молодец 👍

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> После добавления валидационной выборки, сделал также transform на валидации.  </div>

<div class="alert alert-success">
    <b> Комментарий ревьюера v2✔️: </b>
    
Все правильно сделал)
</div>

In [12]:
# Обучим модель логистической регрессии без учета дисбаланса классов и сделаем прогноз.
best_score = 0
best_params = None

for max_iter in range(0, 1000, 10):
    model_lr = LogisticRegression(random_state=12345, max_iter=max_iter, solver='liblinear')
    model_lr.fit(features_train, target_train)
    predicted_valid_lr = model_lr.predict(features_valid)
    predicted_prob_lr = model_lr.predict_proba(features_valid)[:, 1]

    f1 = f1_score(target_valid, predicted_valid_lr)
    auc_roc = roc_auc_score(target_valid, predicted_prob_lr)

    if f1 > best_score:
        best_score = f1
        best_params = {'max_iter': max_iter}

print('Лучшие гиперпараметры:', best_params)
model_lr = LogisticRegression(random_state=12345, solver='liblinear', **best_params)
model_lr.fit(features_train, target_train)
predicted_valid_lr = model_lr.predict(features_valid)
predicted_prob_lr = model_lr.predict_proba(features_valid)[:, 1]
print('F1-оценка лог. рег. без учета дисбаланса классов: {:.2f}'.format(f1_score(target_valid, predicted_valid_lr)))
print('Оценка AUC-ROC лог. рег. без учета дисбаланса классов: {:.2f}'.format(roc_auc_score(target_valid, predicted_prob_lr)))

/opt/conda/lib/python3.9/site-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


Лучшие гиперпараметры: {'max_iter': 10}
F1-оценка лог. рег. без учета дисбаланса классов: 0.28
Оценка AUC-ROC лог. рег. без учета дисбаланса классов: 0.76


<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b>  
    
Вижу, что в работе нет никакого перебора гиперпараметров. Определенно стоит добавить, если хочешь достичь целевой метрики.
</div>

<div class="alert alert-danger">
<b>Комментарий ревьюера ❌:</b> 

AUC-ROC рассчитывается через предсказанные вероятности класса, стоит использовать `model_lg.predict_proba(features_test)`.

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал! </div>

In [13]:
print(classification_report(target_valid, predicted_valid_lr))

              precision    recall  f1-score   support

           0       0.82      0.96      0.89      1447
           1       0.56      0.19      0.28       371

    accuracy                           0.80      1818
   macro avg       0.69      0.58      0.58      1818
weighted avg       0.77      0.80      0.76      1818



*Из отчета о классификации видно, что модель хорошо работает в прогнозировании клиентов, которые не собираются уходить, но имеет низкую эффективность в выявлении клиентов, которые собираются уходить. Поэтому может потребоваться улучшить способность модели идентифицировать ушедших клиентов, чтобы делать более точные прогнозы либо использовать другие модели.* 

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b>  
    
***В нашем случае показатель F1 равен 0.312, что указывает на низкую точность и полноту модели.*** - не факт. Можно отдельно посчитать эти метрики и сделать вывод или использовать `classification_report` (<a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html">ссылка</a>)
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал! Очень удобный инструмент - не знал о нём, спасибо большое! </div>

<div class="alert alert-success">
<b>Комментарий ревьюера ✔️:</b> 

Могу также предложить построить <a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.ConfusionMatrixDisplay.html"> `confusion matrix` </a>.

Матрица ошибок — это таблица, которая позволяет визуализировать эффективность алгоритма классификации путем сравнения прогнозируемого значения целевой переменной с ее фактическим значением. Столбцы матрицы представляют наблюдения в прогнозируемом классе, а строки — наблюдения в фактическом классе (или наоборот).
    
</div>

In [14]:
confusion_matrix(target_valid, predicted_valid_lr)

array([[1393,   54],
       [ 301,   70]])

*В нашем случае матрица ошибок показывает, что модель правильно идентифицировала большое количество отрицательных случаев (клиенты, которые не ушли), но также имеет большое количество ложноотрицательных результатов (клиенты, которые фактически ушли, но, по прогнозам, останутся). Это говорит о том, что модель может не учитывать все важные факторы, влияющие на отток клиентов, и, возможно, ее необходимо улучшить, включив дополнительные признаки, изменив гиперпараметры модели или использовать другую модель.*

In [15]:
# Обучим и проверим классификатор случайного леса без учета дисбаланса.
best_score = 0
best_params = {}

for n_estimators in range(10, 101, 10):
    for max_depth in range(1, 16):
        model_rfc = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth, random_state=12345)
        model_rfc.fit(features_train, target_train)
        predicted_valid_rfc = model_rfc.predict(features_valid)
        predicted_prob_rfc = model_rfc.predict_proba(features_valid)[:, 1]

        f1 = f1_score(target_valid, predicted_valid_rfc)
        auc_roc = roc_auc_score(target_valid, predicted_prob_rfc)

        if f1 > best_score:
            best_score = f1
            best_params = {'n_estimators': n_estimators, 'max_depth': max_depth}

print('Лучшие гиперпараметры:', best_params)
model_rfc = RandomForestClassifier(random_state=12345, **best_params)
model_rfc.fit(features_train, target_train)
predicted_valid_rfc = model_rfc.predict(features_valid)
predicted_prob_rfc = model_rfc.predict_proba(features_valid)[:, 1] 
print('F1-оценка случайного леса без учета дисбаланса классов: {:.2f}'.format(f1_score(target_valid, predicted_valid_rfc)))
print('Оценка AUC-ROC случайного леса без учета дисбаланса классов: {:.2f}'.format(roc_auc_score(target_valid, predicted_prob_rfc)))

Лучшие гиперпараметры: {'n_estimators': 20, 'max_depth': 11}
F1-оценка случайного леса без учета дисбаланса классов: 0.59
Оценка AUC-ROC случайного леса без учета дисбаланса классов: 0.86


In [16]:
print(classification_report(target_valid, predicted_valid_rfc))

              precision    recall  f1-score   support

           0       0.88      0.97      0.92      1447
           1       0.79      0.47      0.59       371

    accuracy                           0.87      1818
   macro avg       0.83      0.72      0.75      1818
weighted avg       0.86      0.87      0.85      1818



In [17]:
confusion_matrix(target_valid, predicted_valid_rfc)

array([[1400,   47],
       [ 197,  174]])

*Заметен рост показателей, по сравнению с логистической регрессией и целевая метрика достигнута, но необходимо проработать дисбаланс классов при помощи библиотеки imblearn чтобы посмотреть улучшится ли результат.*

## Борьба с дисбалансом

Улучшим качество модели с учетом дисбаланса классов. Для этого обучим разные модели и найдем лучшую.

<div class="alert alert-success">
<b>Комментарий ревьюера v2 ✔️:</b> 

Наверное, это в предыдущий раздел? Тут вижу не меняешь выборку.
</div>

In [18]:
# Увеличим выборку класса меньшинства, используя SMOTE.
best_score = 0
best_params = {}

smote = SMOTE(random_state=12345)

for n_estimators in range(10, 101, 10):
    for max_depth in range(1, 16):
        model_rfc = RandomForestClassifier(n_estimators=n_estimators, max_depth=max_depth, random_state=12345)

        features_train_resampled, target_train_resampled = smote.fit_resample(features_train, target_train)
        model_rfc.fit(features_train_resampled, target_train_resampled)

        predicted_valid_rfc = model_rfc.predict(features_valid)
        predicted_valid_rfc_proba = model_rfc.predict_proba(features_valid)[:,1]

        f1 = f1_score(target_valid, predicted_valid_rfc)
        auc_roc = roc_auc_score(target_valid, predicted_valid_rfc_proba)

        if f1 > best_score:
            best_score = f1
            best_params = {'n_estimators': n_estimators, 'max_depth': max_depth}

print('Лучшие гиперпараметры:', best_params)

model_rfc_smote = RandomForestClassifier(random_state=12345, **best_params)
features_train_resampled, target_train_resampled = smote.fit_resample(features_train, target_train)
model_rfc_smote.fit(features_train_resampled, target_train_resampled)

predicted_valid_rfc_smote = model_rfc_smote.predict(features_valid)
predicted_valid_rfc_smote_proba = model_rfc_smote.predict_proba(features_valid)[:,1]

print('F1-оценка случайного леса с учетом дисбаланса классов: {:.2f}'.format(f1_score(target_valid, predicted_valid_rfc_smote)))
print('Оценка AUC-ROC случайного леса с учетом дисбаланса классов: {:.2f}'.format(roc_auc_score(target_valid, predicted_valid_rfc_smote_proba)))

Лучшие гиперпараметры: {'n_estimators': 30, 'max_depth': 15}
F1-оценка случайного леса с учетом дисбаланса классов: 0.60
Оценка AUC-ROC случайного леса с учетом дисбаланса классов: 0.85


<div class="alert alert-danger">
<b>Комментарий ревьюера v2 ❌:</b> 

К сожалению, придется еще чуто доработать проект.
    
Не надо использовать grid search, когда ты меняешь обучающую выборку. У тебя сейчас получается, что ты ее меняешь выборку, на ней же тренируешь и предсказываешь. Происходит утечка данных.
    
Можно запариться, использовать `Kfold`, делить выборку по индексам, выбирать трейн, его увеличивать и потом обучать и предсказывать. Как вариант, еще использовтаь `Pipeline`. Но тему кросс-валидации вы пройдете дальше по курсу. Сейчас советую все же просто сделать, например, обычный цикл.

</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал! </div>

<div class="alert alert-success">
<b> Комментарий ревьюера v3 ✔️: </b>
    
Отлично!
</div>

In [19]:
print(classification_report(target_valid, predicted_valid_rfc_smote))

              precision    recall  f1-score   support

           0       0.90      0.88      0.89      1447
           1       0.58      0.62      0.60       371

    accuracy                           0.83      1818
   macro avg       0.74      0.75      0.75      1818
weighted avg       0.83      0.83      0.83      1818



In [20]:
confusion_matrix(target_valid, predicted_valid_rfc_smote)

array([[1277,  170],
       [ 140,  231]])

- Результаты показывают, что использование SMOTE для устранения дисбаланса классов в обучающих данных - улучшило оценку F1 на валидационной выборке по сравнению с предыдущей моделью без SMOTE.

- Оценка F1 на валидационной выборке составляет 0.60, что лучше, чем оценка F1 предыдущей модели, равная 0.59, что указывает на то, что метод SMOTE улучшил способность модели предсказывать класс меньшинства.

- Оценка AUC-ROC составляет 0.85, что является относительно высоким показателем, указывающим на то, что модель обладает хорошей способностью различать положительные и отрицательные классы.

<div class="alert alert-success">
<b> Комментарий ревьюера ✔️: </b>
    
Для борьбы с дисбалансом часто используется бибилиотека `imblearn`. Особое внимание хотела бы обратить на метод `SMOTE`.
<a href="https://towardsdatascience.com/the-5-most-useful-techniques-to-handle-imbalanced-datasets-6cdba096d55a">Ссылка</a> на статью. Интересная вещь и на собеседованиях часто спрашивают)

</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера ⚠️:</b>  

Давай попробуем еще другие методы работы с дисбалансом, так как показатель целевой метрики не достигнут!
    
</div>

<div class="alert alert-info"> <b>Комментарий студента:</b> Сделал! Отличная библиотека для работы с дисбалансом классов - не знал о ней, на курсе в теоретической части о ней вообще ничего не рассказывали - спасибо большое что поделилась знанием! Благодарю!)) </div>

<div class="alert alert-success">
<b> Комментарий ревьюера v2 ✔️: </b>
    
Ну так нижу вижу результат достигнут, все отлично)
</div>

## Тестирование модели

In [23]:
# Оценим производительность моделей на тестовых данных
predictions_test = model_rfc_smote.predict(features_test)
predicted_test_rfc_smote_proba = model_rfc_smote.predict_proba(features_test)[:,1]

f1 = f1_score(target_test, predictions_test)
auc_roc = roc_auc_score(target_test, predicted_test_rfc_smote_proba)
print("Окончательные результаты тестирования классификатора случайного леса с увеличением выборки с помощью метода SMOTE.")
print("F1: {:.2f}".format(f1))
print("AUC-ROC: {:.2f}".format(auc_roc))

Окончательные результаты тестирования классификатора случайного леса с увеличением выборки с помощью метода SMOTE.
F1: 0.59
AUC-ROC: 0.84


<div class="alert alert-warning">
<b>Комментарий ревьюера v2⚠️:</b>  

Оставлять результаты теста стоит только на модели, выбранной после валидации. 
</div>

**Вывод**

    1. Судя по результатам, классификатор случайного леса с увеличением числа примеров миноритарного класса при помощи метода SMOTE превзошел две другие модели. Показатель F1 достиг целевого показателя (0.59) на тестовой выборке, а показатель AUC-ROC составил 0.84. Эти улучшения предполагают, что метод SMOTE успешно сбалансировал данные и улучшил общую производительность модели.

    2. При сравнении двух моделей, в которых не учитывался дисбаланс, классификатор случайного леса работал лучше, чем модель логистической регрессии, с точки зрения обоих метрик (F1 и AUC-ROC). Это говорит о том, что модель случайного леса может лучше подходить для этого конкретного набора данных.

   В целом - основной вывод такой, что при построении моделей машинного обучения важно учитывать дисбаланс классов. SMOTE является эффективным методом, используемым для устранения дисбаланса классов, и полученные здесь результаты показывают, что он может быть эффективным также для улучшения производительности модели. Однако важно отметить, что результаты могут различаться в зависимости от конкретного набора данных и используемой модели, и может потребоваться попробовать несколько методов, чтобы найти наилучший подход для конкретной проблемы.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*